File: python/app/services/export.py
"""
Export service for handling various export operations
"""
import asyncio
import io
from datetime import datetime
from pathlib import Path
from typing import AsyncIterator, Optional

from app.config import settings
from app.models.order import Order
from app.models.task import ExportTask, TaskStatus
from app.schemas.export import ExportFormat, ExportRequest
from app.schemas.order import OrderQueryParams
from app.services.order import OrderService
from app.services.task import TaskService
from app.utils.csv_writer import CSVWriter
from app.utils.excel_writer import ExcelWriter


class ExportService:
    """Service for export operations"""

    @staticmethod
    async def sync_export(params: ExportRequest) -> tuple[bytes, str, int]:
        """
        Synchronous export - returns file content directly

        Args:
            params: Export request parameters

        Returns:
            Tuple of (file content, filename, file size)
        """
        # Build query params
        query_params = OrderQueryParams(
            order_no=params.order_no,
            user_id=params.user_id,
            order_status=params.order_status,
            start_time=params.start_time,
            end_time=params.end_time,
            min_amount=params.min_amount,
            max_amount=params.max_amount,
            page=1,
            page_size=10000,  # Limit for sync export
        )

        # Get orders
        orders = await OrderService.get_orders_for_export(query_params)

        # Generate file content
        if params.format == ExportFormat.CSV:
            content = await CSVWriter.generate_csv_bytes(orders)
            filename = f"orders_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"
        else:
            content = await ExcelWriter.generate_excel_bytes(orders)
            filename = f"orders_{datetime.now().strftime('%Y%m%d_%H%M%S')}.xlsx"

        return content, filename, len(content)

    @staticmethod
    async def create_async_export_task(
        params: ExportRequest,
        api_key: Optional[str] = None,
    ) -> ExportTask:
        """
        Create async export task

        Args:
            params: Export request parameters
            api_key: API Key for authentication

        Returns:
            Created export task
        """
        # Build query params dict
        query_params = {
            "order_no": params.order_no,
            "user_id": params.user_id,
            "order_status": params.order_status,
            "start_time": params.start_time.isoformat() if params.start_time else None,
            "end_time": params.end_time.isoformat() if params.end_time else None,
            "min_amount": float(params.min_amount) if params.min_amount else None,
            "max_amount": float(params.max_amount) if params.max_amount else None,
        }

        # Create task
        task = await TaskService.create_task(
            export_format=params.format,
            query_params=query_params,
            api_key=api_key,
        )

        return task

    @staticmethod
    async def process_export_task(task_id: str) -> None:
        """
        Process export task in background

        Args:
            task_id: Task ID to process
        """
        # Get task
        task = await TaskService.get_task_by_id(task_id)
        if not task:
            return

        try:
            # Mark task as started
            await TaskService.start_task(task_id)

            # Parse query params
            query_params_dict = task.query_params or {}
            query_params = OrderQueryParams(
                order_no=query_params_dict.get("order_no"),
                user_id=query_params_dict.get("user_id"),
                order_status=query_params_dict.get("order_status"),
                start_time=(
                    datetime.fromisoformat(query_params_dict["start_time"])
                    if query_params_dict.get("start_time")
                    else None
                ),
                end_time=(
                    datetime.fromisoformat(query_params_dict["end_time"])
                    if query_params_dict.get("end_time")
                    else None
                ),
                min_amount=query_params_dict.get("min_amount"),
                max_amount=query_params_dict.get("max_amount"),
                page=1,
                page_size=1000000,  # Large limit for export
            )

            # Get orders
            orders = await OrderService.get_orders_for_export(query_params)
            total_count = len(orders)

            # Update total count
            await TaskService.update_task_progress(task_id, 0, total_count)

            # Create export directory
            export_dir = Path(settings.EXPORT_DIR)
            export_dir.mkdir(parents=True, exist_ok=True)

            # Generate file
            filename = f"orders_{task_id}.{task.export_format}"
            file_path = export_dir / filename

            # Write file
            if task.export_format == ExportFormat.CSV.value:
                count = await CSVWriter.write_to_file(orders, file_path)
            else:
                count = await ExcelWriter.write_to_file(orders, file_path)

            # Update progress
            await TaskService.update_task_progress(task_id, count, total_count)

            # Mark task as completed
            file_size = file_path.stat().st_size
            await TaskService.complete_task(task_id, str(file_path), file_size)

        except Exception as e:
            # Mark task as failed
            await TaskService.fail_task(task_id, str(e))
            raise

    @staticmethod
    async def stream_export(params: ExportRequest) -> AsyncIterator[bytes]:
        """
        Stream export data

        Args:
            params: Export request parameters

        Yields:
            File content chunks
        """
        # Build query params
        query_params = OrderQueryParams(
            order_no=params.order_no,
            user_id=params.user_id,
            order_status=params.order_status,
            start_time=params.start_time,
            end_time=params.end_time,
            min_amount=params.min_amount,
            max_amount=params.max_amount,
            page=1,
            page_size=1000000,  # Large limit for export
        )

        # Get orders
        orders = await OrderService.get_orders_for_export(query_params)

        # Stream CSV data
        if params.format == ExportFormat.CSV:
            async for chunk in CSVWriter.stream_csv(orders):
                yield chunk.encode("utf-8-sig")
        else:
            # For Excel, we need to generate the entire file first
            # Then stream it in chunks
            content = await ExcelWriter.generate_excel_bytes(orders)
            chunk_size = 8192
            for i in range(0, len(content), chunk_size):
                yield content[i : i + chunk_size]
                await asyncio.sleep(0)

    @staticmethod
    async def get_export_file(token: str) -> Optional[tuple[Path, str]]:
        """
        Get export file by download token

        Args:
            token: Download token

        Returns:
            Tuple of (file path, filename) or None
        """
        # Get task
        task = await TaskService.get_task_by_token(token)
        if not task or not task.is_completed:
            return None

        # Check if file exists
        if not task.file_path:
            return None

        file_path = Path(task.file_path)
        if not file_path.exists():
            return None

        # Generate filename
        filename = f"orders_{task.created_at.strftime('%Y%m%d_%H%M%S')}.{task.export_format}"

        return file_path, filename

File: rust/src/services/export.rs
use std::fs::{self, File};
use std::io::Write;
use std::path::PathBuf;
use std::sync::Arc;
use uuid::Uuid;

use crate::db::DbPool;
use crate::models::{ExportRequest, ExportTask, Order, OrderQueryParams, TaskStatus, SSEEventData};
use crate::services::order::OrderService;
use crate::services::task::TaskManager;
use crate::utils::csv::write_csv;
use crate::utils::excel::write_excel;

/// 导出服务
pub struct ExportService;

impl ExportService {
    /// 同步导出（直接返回文件）
    pub fn export_sync(
        pool: &DbPool,
        export_dir: &str,
        request: ExportRequest,
    ) -> Result<(PathBuf, String), String> {
        // 创建导出目录
        fs::create_dir_all(export_dir)
            .map_err(|e| format!("Failed to create export directory: {}", e))?;

        let mut conn = pool.get()
            .map_err(|e| format!("Failed to get database connection: {}", e))?;

        // 构建查询参数
        let params = OrderQueryParams {
            page: None,
            page_size: None,
            start_time: request.start_time,
            end_time: request.end_time,
            order_status: request.order_status,
            min_amount: request.min_amount,
            max_amount: request.max_amount,
            user_id: request.user_id,
            order_no: request.order_no,
        };

        // 查询所有订单
        let total = OrderService::count_orders(&mut conn, &params)?;
        let orders = OrderService::query_orders_batch(&mut conn, &params, 0, total)?;

        // 生成文件名
        let format = request.format.unwrap_or_else(|| "csv".to_string());
        let filename = format!(
            "export_{}.{}",
            chrono::Local::now().format("%Y%m%d_%H%M%S"),
            format
        );
        let filepath = PathBuf::from(export_dir).join(&filename);

        // 写入文件
        match format.as_str() {
            "csv" => write_csv(&filepath, &orders)?,
            "xlsx" | "excel" => write_excel(&filepath, &orders)?,
            _ => return Err(format!("Unsupported format: {}", format)),
        }

        Ok((filepath, format))
    }

    /// 异步导出（后台任务）
    pub fn export_async(
        pool: DbPool,
        export_dir: String,
        request: ExportRequest,
        task_manager: Arc<TaskManager>,
    ) -> ExportTask {
        let format = request.format.clone().unwrap_or_else(|| "csv".to_string());
        let task = task_manager.create_task(format.clone());

        let task_id = task.task_id;
        let task_manager_clone = task_manager.clone();

        // 启动后台任务
        tokio::spawn(async move {
            // 创建导出目录
            if let Err(e) = fs::create_dir_all(&export_dir) {
                let mut task = task_manager_clone.get_task(&task_id).unwrap();
                task.status = TaskStatus::Failed;
                task.error_message = Some(format!("Failed to create export directory: {}", e));
                task_manager_clone.update_task(&task);
                return;
            }

            // 获取数据库连接
            let mut conn = match pool.get() {
                Ok(c) => c,
                Err(e) => {
                    let mut task = task_manager_clone.get_task(&task_id).unwrap();
                    task.status = TaskStatus::Failed;
                    task.error_message = Some(format!("Failed to get database connection: {}", e));
                    task_manager_clone.update_task(&task);
                    return;
                }
            };

            // 构建查询参数
            let params = OrderQueryParams {
                page: None,
                page_size: None,
                start_time: request.start_time,
                end_time: request.end_time,
                order_status: request.order_status,
                min_amount: request.min_amount,
                max_amount: request.max_amount,
                user_id: request.user_id,
                order_no: request.order_no,
            };

            // 更新任务状态为处理中
            let mut task = task_manager_clone.get_task(&task_id).unwrap();
            task.status = TaskStatus::Processing;
            task_manager_clone.update_task(&task);

            // 查询总数
            let total = match OrderService::count_orders(&mut conn, &params) {
                Ok(t) => t as u32,
                Err(e) => {
                    task.status = TaskStatus::Failed;
                    task.error_message = Some(e);
                    task_manager_clone.update_task(&task);
                    return;
                }
            };

            task.total = total;
            task_manager_clone.update_task(&task);

            // 分批查询并写入文件
            let batch_size = 1000;
            let mut all_orders = Vec::new();

            for offset in (0..total as i64).step_by(batch_size as usize) {
                let orders = match OrderService::query_orders_batch(
                    &mut conn,
                    &params,
                    offset,
                    batch_size,
                ) {
                    Ok(o) => o,
                    Err(e) => {
                        task.status = TaskStatus::Failed;
                        task.error_message = Some(e);
                        task_manager_clone.update_task(&task);
                        return;
                    }
                };

                all_orders.extend(orders);
                task.progress = all_orders.len() as u32;
                task_manager_clone.update_task(&task);
            }

            // 生成文件
            let filename = format!(
                "export_{}.{}",
                chrono::Local::now().format("%Y%m%d_%H%M%S"),
                format
            );
            let filepath = PathBuf::from(&export_dir).join(&filename);

            let write_result = match format.as_str() {
                "csv" => write_csv(&filepath, &all_orders),
                "xlsx" | "excel" => write_excel(&filepath, &all_orders),
                _ => Err(format!("Unsupported format: {}", format)),
            };

            match write_result {
                Ok(_) => {
                    task.status = TaskStatus::Completed;
                    task.file_path = Some(filepath.to_string_lossy().to_string());
                    task.download_token = Some(Uuid::new_v4().to_string());
                    task.completed_at = Some(chrono::Utc::now());
                }
                Err(e) => {
                    task.status = TaskStatus::Failed;
                    task.error_message = Some(e);
                }
            }

            task_manager_clone.update_task(&task);
        });

        task
    }
}
